src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -61,20 +64,35 @@ public void run() {
    */
   private void runNodes(List<T> freeNodes) {
     List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
+    CountDownLatch latch = null;
+    if (useLatches && runnables.size() > 1) {
+      latch = new CountDownLatch(runnables.size());
+    }
+    boolean latchUsed = false;
     for (IWorker<T> r : runnables) {
       m_activeRunnables.add(r);
+      if (latch != null && (r instanceof TestMethodWorker)) {
+        //Being conservative here and setting this only for test method workers
+        r.setLatch(latch);
+        latchUsed = true;
+      }
       ppp("Added to active runnable");
       setStatus(r, Status.RUNNING);
       ppp("Executing: " + r);
       try {
         execute(r);
-//        if (m_threadCount > 1) execute(r);
-//        else r.run();
       }
       catch(Exception ex) {
         ex.printStackTrace();
       }
     }
+    if (latch != null && latchUsed) {
+      try {
+        latch.await(10, TimeUnit.SECONDS);
```

Comment:
No we are not waiting for the end of all runnables. We are waiting for the completion of execution of all workers that are part of the freenodes. So for e.g., I have 3 methods for each priority (1, 2, 3), the latch would wait for all the 3 methods of priority 1 to run to completion, before it starts picking up methods of priority 2 and running them. Absence of this behavior was what was causing the tests to fail. 

awaitTermination, I believe is being down outside at `TestRunner` level, wherein the executor is being kicked off. This is more to do with kind of bringing in a discipline amongst tests that execute the `IWorker` variants for a given set of nodes. I hope I managed to explain it :)